components/script/dom/document.rs

```diff
@@ -4082,6 +4090,31 @@ impl Document {
     pub(crate) fn set_declarative_refresh(&self, refresh: DeclarativeRefresh) {
         *self.declarative_refresh.borrow_mut() = Some(refresh);
     }
+
+    /// <https://html.spec.whatwg.org/multipage/interaction.html#visibility-state>
+    fn update_visibility_state(&self, visibility_state: DocumentVisibilityState) {
+        // Step 1
+        if self.visibility_state.get() == visibility_state {
+            return;
+        }
+        // Step 2
+        self.visibility_state.set(visibility_state);
+        // Step 3 TODO get timestamp
+        let entry = VisibilityStateEntry::new(
+            &self.global(),
+            visibility_state,
+            *self.global().performance().Now(),
+        );
+        self.window
+            .Performance()
+            .queue_entry(entry.upcast::<PerformanceEntry>());
+        // Step 4 TODO ScreenOrientation hasn't implemented yet
+        // Step 5 TODO ViewTransition hasn't implemented yet
+        // Step 6 TODO This step is too vague
```

Comment:
@gterzian, I opened a draft for now because I want to know if we could skip these.
It seems Servo hasn't implemented the parts required by steps 4 & 5. While step 6 is so vague that it even has a comment section discussing it.

Reply:
Yes on the TODO for 4 and 5, re 6 this is more like a hook for other specs to defines steps, so as those other specs are implemented they could run their visibility steps here. For example https://www.w3.org/TR/gamepad/#handling-visibility-change (even though they don't mention these steps direclty but I think that's the idea) cc @msub2

Reply:
Does this mean I need to add the implementation of gamepad's step here?

Reply:
As long as page visibility is being handled properly I can implement Gamepad-specific behavior in a future PR

Reply:
Alright, I updated the description text and I'll move this draft to PR soon.